packages/fxa-content-server/app/scripts/lib/glean/index.ts

```diff
@@ -2,32 +2,90 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+import Glean from '@mozilla/glean/web';
+
+import { accountsEvents } from './pings';
+import { name } from './event';
+import { uid } from './account';
+import { oauthClientId, service } from './relyingParty';
+import { deviceType, entrypoint, flowId } from './session';
+import * as utm from './utm';
+
 export type GleanMetricsConfig = {
   enabled: boolean;
   applicationId: string;
   uploadEnabled: boolean;
   appDisplayVersion: string;
   channel: string;
   serverEndpoint: string;
+  logPings: boolean;
+  debugViewTag: string;
 };
 
-import Glean from '@mozilla/glean/web';
-import * as registration from './registration';
+export type GleanMetricsContext = {
+  metrics: any;
+  relier: any;
+  user: any;
+  userAgent: any;
+};
+
+let gleanEnabled = false;
+let gleanMetricsConext;
+
+const populateProperties = () => {
+  const flowEventMetadata = gleanMetricsConext.metrics.getFlowEventMetadata();
+
+  // TODO when sending metrics for authenticated accounts
+  uid.set('');
+
+  oauthClientId.set(gleanMetricsConext.relier.get('clientId') || '');
+  service.set(gleanMetricsConext.relier.get('service') || '');
+
+  deviceType.set(gleanMetricsConext.userAgent.genericDeviceType() || '');
+  entrypoint.set(flowEventMetadata.entrypoint || '');
+  flowId.set(flowEventMetadata.flowId || '');
+
+  utm.campaign.set(flowEventMetadata.utmCampaign || '');
+  utm.content.set(flowEventMetadata.utmContent || '');
+  utm.medium.set(flowEventMetadata.utmMedium || '');
+  utm.source.set(flowEventMetadata.utmSource || '');
+  utm.term.set(flowEventMetadata.utmTerm || '');
+};
+
+const createEventFn = (eventName) => () => {
+  if (!gleanEnabled) {
+    return;
+  }
+
+  name.set(eventName);
+  populateProperties();
+  accountsEvents.submit();
+};
 
 export const GleanMetrics = {
-  initialize: (config: GleanMetricsConfig) => {
+  initialize: (config: GleanMetricsConfig, context: GleanMetricsContext) => {
     if (config.enabled) {
       Glean.initialize(config.applicationId, config.uploadEnabled, {
         appDisplayVersion: config.appDisplayVersion,
         channel: config.channel,
         serverEndpoint: config.serverEndpoint,
-        // Glean does not offer direct control over when metrics are submitted;
-        // this ensures that events are submitted.
+        // Glean does not offer direct control over when metrics are uploaded;
+        // this ensures that events are uploaded.
         maxEvents: 1,
       });
+      Glean.setLogPings(config.logPings);
+      if (config.debugViewTag) {
+        Glean.setDebugViewTag(config.debugViewTag);
+      }
+
+      gleanEnabled = config.enabled;
```

Comment:
I think there might be an edge case here. If the config.enabled is false, this line will never get hit, which means the following could happen:

```
// Turns gleanEnabled true
GleanMetrics.initialize({ enabled: true, ...config1}, context1);

// Would submit as expected
GleanMetrics.registration.view();

// Try disabling glean... Not typical, but I can see someone doing this.
GleanMetrics.initialize({ enabled: false, ...config2}, context2):

// Here's the edge case. I think this would still submit an event using the 
// config1 and context1 states. 
GleanMetrics.registration.view();
```

Reply:
😳 Good catch.  Thanks!